src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -248,25 +248,16 @@ private bool TypeIsCollectionOfNumbersWithInternalConverter()
             Type? elementType = ConverterBase.ElementType;
             Debug.Assert(elementType != null);
 
-            elementType = Nullable.GetUnderlyingType(elementType) ?? elementType;
-
-            if (elementType == typeof(byte) ||
-                elementType == typeof(decimal) ||
-                elementType == typeof(double) ||
-                elementType == typeof(short) ||
-                elementType == typeof(int) ||
-                elementType == typeof(long) ||
-                elementType == typeof(sbyte) ||
-                elementType == typeof(float) ||
-                elementType == typeof(ushort) ||
-                elementType == typeof(uint) ||
-                elementType == typeof(ulong) ||
-                elementType == JsonClassInfo.ObjectType)
+            try
```

Comment:
Previous logic did not verify that the converter for the collection element is internal, meaning that a call to `ReadNumberWithCustomHandling` would [throw](https://github.com/dotnet/runtime/blob/e13f72bfe26ef243fa96e67ec3ee5aa29c294c54/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L511-L512).

The write method is fine because we have an additional check that the [converter is internal and a number type](https://github.com/dotnet/runtime/blob/e13f72bfe26ef243fa96e67ec3ee5aa29c294c54/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L359). The write method has this check to support boxed numbers on serialization (but not call the new virtual method for incompatible types like `DateTime`). This additional check [is not done](https://github.com/dotnet/runtime/blob/e13f72bfe26ef243fa96e67ec3ee5aa29c294c54/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L182) on deserialization as we simply create a new `JsonElement` of `JsonValueKind.String`.